compiler/cpp/test/ReaderTest.cpp

```diff
@@ -0,0 +1,234 @@
+        reader.expect('u');
+        SUCCEED();
+    } catch (UnexpectedCharException const & err) {
+        FAIL() << "Expected UnexpectedCharException got thrown";
+    }
+}
+
+TEST(ReaderTests, ExpectFailure) {
+    // Initialize
+    Reader reader = Reader("File input", 5);
+
+    // Test
+    reader.advance();
+    reader.expect('i');
+    reader.expect('l');
+    reader.expect('e');
+    reader.advance();
+    try {
+        reader.expect('n');
+        FAIL() << "Expected UnexpectedCharException";
+    } catch (UnexpectedCharException const & err) {
+        EXPECT_EQ(err.what(),std::string("Expected 'n', but got 'i' at L1 C6"));
+    }
+}
+
+// --------------------------------------------------- Expect Multiple -------------------------------------------------
+
+TEST(ReaderTests, ExpectMultipleSuccess) {
+    // Initialize
+    Reader reader = Reader("File input", 5);
+
+    // Test
+    try {
+        reader.advance();
+        reader.expectMultiple("ile");
+        reader.advance();
+        reader.expectMultiple("in");
+        reader.advance();
+        reader.expectMultiple("ut");
+        SUCCEED();
+    } catch (UnexpectedCharException const & err) {
+        FAIL() << "Expected UnexpectedCharException got thrown";
+    }
+}
+
+TEST(ReaderTests, ExpectMultipleFailure) {
+    // Initialize
+    Reader reader = Reader("File input", 5);
+
+    // Test
+    reader.advance();
+    reader.expectMultiple("ile");
+    reader.advance();
+    try {
+        reader.expectMultiple("data");
+        FAIL() << "Expected UnexpectedCharException";
+    } catch (UnexpectedCharException const & err) {
+        EXPECT_EQ(err.what(),std::string("Expected 'd', but got 'i' at L1 C6"));
+    }
+}
+
+// ------------------------------------------------------- Code Pos ----------------------------------------------------
+
+TEST(ReaderTests, CodePos) {
+    // Initialize
+    Reader reader = Reader("File input", 5);
+
+    // Test
+    EXPECT_EQ(reader.getLineNum(), 1);
+    EXPECT_EQ(reader.getColNum(), 1);
+    reader.advance();
+    reader.advance();
+    reader.advance();
+    EXPECT_EQ(reader.getLineNum(), 1);
+    EXPECT_EQ(reader.getColNum(), 4);
+}
+
+TEST(ReaderTests, CodePosLineBreaks) {
+    // Initialize
+    Reader reader = Reader("First\nSecond\n\nFourth", 5);
+
+    // Test
+    EXPECT_EQ(reader.getLineNum(), 1);
+    EXPECT_EQ(reader.getColNum(), 1);
+    reader.expectMultiple("First");
+    reader.advance(); // Consume line break
+    EXPECT_EQ(reader.getLineNum(), 2);
+    EXPECT_EQ(reader.getColNum(), 1);
+    reader.expectMultiple("Second");
+    reader.advance(); // Consume line break
+    reader.advance(); // Consume line break
+    reader.expectMultiple("Fou");
+    EXPECT_EQ(reader.getLineNum(), 4);
+    EXPECT_EQ(reader.getColNum(), 4);
+}
+
+TEST(ReaderTests, CodePosEmptyFile) {
+    // Initialize
+    Reader reader = Reader("", 3);
+
+    // Test
+    EXPECT_EQ(reader.getLineNum(), 1);
+    EXPECT_EQ(reader.getColNum(), 0);
```

Comment:
Why should `getColNum` return `0` at the beginning? In most text editors (including IntelliJ), we start at `1:1`.

Reply:
This is right. I've added a check for this so whenever the initial `advance` call fails (due to an empty input string) it sets the colNum to 1 instread of 0.

Reply:
Ok, I don't have to do this since as you stated in another comment, I use EOF as filler --> see the explanation in the other comment